src/engine.rs

```diff
@@ -151,9 +156,12 @@ where
                             .as_secs_f64(),
                     );
                     sync |= writer.sync;
-                    let log_batch = writer.get_payload();
+                    let log_batch = writer.get_mut_payload();
                     let res = if !log_batch.is_empty() {
-                        // @TODO(lucasliang): bind `Version` to each `LogBatch`
+                        // Signs a checksum, so-called `signature`, into the LogBatch.
+                        let (format_version, file_id) =
+                            self.pipe_log.fetch_active_file(LogQueue::Append);
+                        log_batch.sign_checksum(format_version, file_id);
```

Comment:
How about moving this into pipe log? Pipe implementation doesn't guarantee the active file will stay unchanged during `append`.

Reply:
Let me have a look.

Reply:
Here, from my point of view, the reason i did the implementation as this was based on:
>* The caller, who did dump batch of write_data into the `active file`, should only be the leader among writers in `write_barrier`. That is, the `active file` could only be written by one caller(thread) at one time. Also, the manually `maybe_sync` is only called by the leader writer, which is same to `rotate`.
>* The op of `fn sign_checksum` just require the `FileId` of the `active file`. Only when `rotate` was triggered by the leader writer, could `FileId` be updated. And as the previous statement shown, the leader would not `rotate` the `active file` in `append`.
>* Meanwhile, the original design of `append` in `Pipe` is just wanna accept `a serial bytes of data` and dump them into the `active file`. And the encryption on data should be done by LogBatch before we pass it to `Pipe`.

Did i miss something important for this? Or i misunderstand something else?

Reply:
Pipe only guarantees fsync is called in `maybe_sync`, the rotation is implementation details. It's very possible that another implementation will rotate file inside `append`.

Reply:
I agree pipe should only care about raw bytes. I suggest document this contract in pipe trait (which is active_file_id stays unchanged during append call).

Reply:
So, could i keep the design here with the agreement that the `append` call would not update the `active file id`?
And we can truely record the contract with annotations on `Pipe::append` to avoid that other implementations would disobey the rule.